### PR TITLE
Enhancement: use apt only when needed docker-entrypoint.sh

### PR DESCRIPTION
**Pull request by gawa971**
## Proposed change

Currently, when we start paperless container with the env var `PAPERLESS_OCR_LANGUAGES` set with some languages packs, the command `apt-get update` runs **every single time**. Including when it's not needed, in the cases where:
* tesseract packages are detected as "installed" (already exist in the image)
* no new package needs to be installed

I assume most users set the env var `PAPERLESS_OCR_LANGUAGES` once and never touch it again. Some people however need to unset it (in order to match [this condition](https://github.com/paperless-ngx/paperless-ngx/blob/4146b140d3ed5fa4dccb9c01879e8420f6743a5f/docker/docker-entrypoint.sh#L121-L124)) to make sure no network or root operation is done, in the following cases:
* if we run paperless fully rootless (as documented [here](https://docs.paperless-ngx.com/configuration/#PAPERLESS_OCR_LANGUAGES))
* if we run paperless in a restricted network or behind a strict firewall blocking egress port 80

With this change, leaving `PAPERLESS_OCR_LANGUAGES` set is no longer an issue, because `apt` will try to run only when necessary, when it will detect that a new/missing tesseract language package needs to be installed.

Pros:
* Optimization, faster boot time for everyone (apt-get update is fast, but it's really still a waste of time and bandwidth and disk IO to run this where it's not necessary)
* Remove a dependency on apt servers for booting (currently, if apt servers are down or not reachable -> paperless won't start)
* Bonus: might make it slightly easier for some advanced configurations (rootless, strict firewalls)

Cons:
* a slightly more complex bash script

## Type of change

New logic: first, we construct a list of tesseract packages to install, by checking with `dpkg` what package is not installed yet. `dpkg` works as an unprivileged user and offline. We don't need an up to date apt cache for this step.

Then, IF we have some new packages to install, we update the apt cache (`apt-get update`) only once (like before), and use it in the for loop for each package (`apt-cache show`, `apt-get install`).

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

The change does not introduce a breaking change. It will simply make all images boot slightly faster. It's just a small optimization.

This is a minor change to a bash script so most of the following items don't apply (no bash test coverage, no documentation to change...)

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [ ] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

File: docker/docker-entrypoint.sh [hunk 1]
```diff
@@ -122,27 +122,37 @@ install_languages() {
...
+			to_install+=($pkg)
 		fi
+	done
 
-		if ! apt-cache show "$pkg" &>/dev/null; then
-			echo "Package $pkg not found! :("
-			continue
-		fi
+	# Use apt only when we install packages
+	if [ ${#to_install[@]} > 0 ]; then
```
Comment (tribut) [inline docker/docker-entrypoint.sh:139]:
```suggestion
	if [ ${#to_install[@]} -gt 0 ]; then
```

Your version appears to work in bash, but I don't think it's posix. So it should be `[[ ]]` or `-gt`.

**Review by tribut — APPROVED** [review]
For consistency, we should maybe initialize `to_install=()` somewhere, even though we currently don't use `set -u`?

Otherwise this looks good to me.

File: docker/docker-entrypoint.sh [hunk 2]
```diff
@@ -122,27 +122,37 @@ install_languages() {
...
 
+	# Build list of packages to install
 	for lang in "${langs[@]}"; do
 		pkg="tesseract-ocr-$lang"
 
 		if dpkg --status "$pkg" &>/dev/null; then
 			echo "Package $pkg already installed!"
 			continue
+		else
+			to_install+=($pkg)
```
Comment (tribut) [inline docker/docker-entrypoint.sh:134]:
```suggestion
			to_install+=("$pkg")
```

Probably won't matter here, but for consistency: This prevents word splitting/globbing.

**Comment by shamoon** [comment]
Ha, our pre-commit hooks agree with you tribut.

I really do wish we could make the checkboxes in the PR template compulsory.

File: docker/docker-entrypoint.sh (hunk 1 above)
Comment (gawa971) [reply docker/docker-entrypoint.sh]: good (and critical) catch! :see_no_evil:

File: docker/docker-entrypoint.sh (hunk 2 above)
Comment (gawa971) [reply docker/docker-entrypoint.sh]: Done.

**Comment by gawa971** [comment]
> I really do wish we could make the checkboxes in the PR template compulsory.

You'll notice that was the only box I didn't check :sweat_smile: I was keeping that for last, as I wanted to look into it more closely (I was not familiar with the pre-commit python tool). I didn't expect it to have a spellcheck linter. My bad!

I set up the pre-commit now, and spellcheck seems happy on my side.

**Comment by shamoon** [comment]
*shellcheck

All good, thanks.

**Pull request merged**